src/rgw/rgw_bucket.h

```diff
@@ -496,24 +501,65 @@ struct RGWDataChangesLogMarker {
   RGWDataChangesLogMarker() = default;
 };
 
+class RGWDataChangesBE {
+protected:
+  CephContext* const cct;
+private:
+  std::string prefix;
+  static std::string_view get_prefix(CephContext* cct) {
+    std::string_view prefix = cct->_conf->rgw_data_log_obj_prefix;
+    if (prefix.empty()) {
+      prefix = "data_log"sv;
+    }
+    return prefix;
+  }
+public:
+  using entries = std::variant<std::list<cls_log_entry>,
+			       std::vector<ceph::buffer::list>>;
+
+  RGWDataChangesBE(CephContext* const cct)
+    : cct(cct), prefix(get_prefix(cct)) {}
+  virtual ~RGWDataChangesBE() = default;
+
+  static std::string get_oid(CephContext* cct, int i) {
+    return fmt::format("{}.{}", get_prefix(cct), i);
+  }
+  std::string get_oid(int i) {
+    return fmt::format("{}.{}", prefix, i);
+  }
+
+  virtual entries new_entries() = 0;
+  virtual void prepare(ceph::real_time now,
+		       const std::string& key,
+		       ceph::buffer::list&& entry,
+		       entries& out) = 0;
+  virtual int push(int index, entries&& items) = 0;
```

Comment:
oh, now I see why two of everything;  could have used a variant type?  (not that I want you to actually change anything cosmetic if this is ready to merge)